projects/kbd-2.6.4/docs/man/man1/kbdinfo.1
.TH KBDINFO "1" "June 2019" "kbd"
.SH NAME
kbdinfo \- read information about keyboard state
.SH SYNOPSIS
.B kbdinfo
[\fI\,options\/\fR] \fI\,getmode \/\fR[\fI\,text|graphics\/\fR]
.br
.B kbdinfo
[\fI\,options\/\fR] \fI\,gkbmode \/\fR[\fI\,raw|xlate|mediumraw|unicode\/\fR]
.br
.B kbdinfo
[\fI\,options\/\fR] \fI\,gkbmeta \/\fR[\fI\,metabit|escprefix\/\fR]
.br
.B kbdinfo
[\fI\,options\/\fR] \fI\,gkbled  \/\fR[\fI\,scrolllock|numlock|capslock\/\fR]
.SH DESCRIPTION
The utility allows you to read and check various parameters of the keyboard and
virtual console.
.TP 9
.B getmode
Get or check virtual console mode.
.TP 9
.B gkbmode
Gets current keyboard mode.
.LP
.RS
.TP 12
.B raw
Raw (scancode) mode. These are the raw codes generated by the keyboard.
.TP 12
.B mediumraw
Medium raw (scancode) mode. This is extended medium raw mode, with keys above
127 encoded as 0, high 7 bits, low 7 bits, with the 0 bearing the 'up' flag if
needed. 0 is reserved, so this shouldn't interfere with anything else. The two
bytes after 0 will always have the up flag set not to interfere with older
applications. This allows for 16384 different keycodes, which should be enough.
.TP 12
.B xlate
Translate keycodes using keymap. These are the codes generated via the current
keysym mapping.
.TP 12
.B unicode
Unicode mode.
.RE
.LP
.TP 9
.B gkbmeta
Gets meta key handling mode.
.LP
.RS
.TP 12
.B escprefix
Specifies if pressing the meta (alt) key generates an ESC (\\033) prefix followed by the keysym.
.TP 12
.B metabit
The keysym marked with the high bit set.
.RE
.LP

.TP 9
.B gkbled
Get keyboard flags CapsLock, NumLock, ScrollLock (not lights).
.LP
.RS
.TP 12
.B scrolllock
The scroll lock is down.
.TP 12
.B numlock
The num lock is down.
.TP 12
.B capslock
The caps lock is down.
.RE
.LP
.SH OPTIONS
.TP
\fB\-C\fR, \fB\-\-console\fR=\fI\,DEV\/\fR
the console device to be used;
.TP
\fB\-V\fR, \fB\-\-version\fR
print version number;
.TP
\fB\-h\fR, \fB\-\-help\fR
print this usage message.
.SH AUTHORS
Written by Alexey Gladkov.
.SH "SEE ALSO"
.BR kbdrate (1),
.BR setleds (1),
.BR kbd_mode (1)


projects/kbd-2.6.4/src/kbdinfo.c
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/kd.h>
#include <getopt.h>
#include <sysexits.h>

#include "libcommon.h"

static const char *action = NULL;
static const char *value  = NULL;

static void __attribute__((noreturn))
usage(int rc, const struct kbd_help *options)
{
	const struct kbd_help *h;

	fprintf(stderr,
			_("Usage: %1$s [option...] getmode [text|graphics]\n"
			  "   or: %1$s [option...] gkbmode [raw|xlate|mediumraw|unicode]\n"
			  "   or: %1$s [option...] gkbmeta [metabit|escprefix]\n"
			  "   or: %1$s [option...] gkbled  [scrolllock|numlock|capslock]\n"),
			get_progname());
	fprintf(stderr, "\n");
	fprintf(stderr, _(
				"The utility allows to read and check various parameters\n"
				"of the keyboard and virtual console.\n"));

	if (options) {
		int max = 0;

		fprintf(stderr, "\n");
		fprintf(stderr, _("Options:"));
		fprintf(stderr, "\n");

		for (h = options; h && h->opts; h++) {
			int len = (int) strlen(h->opts);
			if (max < len)
				max = len;
		}
		max += 2;

		for (h = options; h && h->opts; h++)
			fprintf(stderr, "  %-*s %s\n", max, h->opts, h->desc);
	}

	fprintf(stderr, "\n");
	fprintf(stderr, _("Report bugs to authors.\n"));
	fprintf(stderr, "\n");

	exit(rc);
}

static int
answer(const char *ans)
{
	if (value)
		return strcasecmp(value, ans) ? EXIT_FAILURE : EXIT_SUCCESS;

	printf("%s\n", ans);
	return EXIT_SUCCESS;
}

int main(int argc, char **argv)
{
	int fd, mode, c;
	int rc = EXIT_FAILURE;
	char flags;
	const char *console = NULL;

	set_progname(argv[0]);
	setuplocale();

	const char *short_opts = "C:hV";
	const struct option long_opts[] = {
		{ "console", required_argument, NULL, 'C' },
		{ "help",    no_argument,       NULL, 'h' },
		{ "version", no_argument,       NULL, 'V' },
		{ NULL,      0,                 NULL,  0  }
	};
	const struct kbd_help opthelp[] = {
		{ "-C, --console=DEV", _("the console device to be used.") },
		{ "-V, --version",     _("print version number.")     },
		{ "-h, --help",        _("print this usage message.") },
		{ NULL, NULL }
	};

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
			case 'C':
				if (optarg == NULL || optarg[0] == '\0')
					usage(EX_USAGE, opthelp);
				console = optarg;
				break;
			case 'V':
				print_version_and_exit();
				break;
			case 'h':
				usage(EXIT_SUCCESS, opthelp);
				break;
			case '?':
				usage(EX_USAGE, opthelp);
				break;
		}
	}

	if (optind == argc) {
		kbd_error(EXIT_FAILURE, 0, _("Not enough arguments."));
	}

	action = argv[optind++];

	if (optind < argc)
		value = argv[optind++];

	if ((fd = getfd(console)) < 0)
		kbd_error(EXIT_FAILURE, 0, _("Couldn't get a file descriptor referring to the console."));

	if (!strcasecmp("GETMODE", action)) {
		if (ioctl(fd, KDGETMODE, &mode) == -1)
			kbd_error(EXIT_FAILURE, errno, _("Unable to read console mode"));

		switch (mode) {
			case KD_TEXT:
				rc = answer("text");
				break;
			case KD_GRAPHICS:
				rc = answer("graphics");
				break;
		}

	} else if (!strcasecmp("GKBMODE", action)) {
		if (ioctl(fd, KDGKBMODE, &mode) == -1)
			kbd_error(EXIT_FAILURE, errno, _("Unable to read keyboard mode"));

		switch (mode) {
			case K_RAW:
				rc = answer("raw");
				break;
			case K_XLATE:
				rc = answer("xlate");
				break;
			case K_MEDIUMRAW:
				rc = answer("mediumraw");
				break;
			case K_UNICODE:
				rc = answer("unicode");
				break;
		}

	} else if (!strcasecmp("GKBMETA", action)) {
		if (ioctl(fd, KDGKBMETA, &mode) == -1)
			kbd_error(EXIT_FAILURE, errno, _("Unable to read meta key handling mode"));

		switch (mode) {
			case K_METABIT:
				rc = answer("metabit");
				break;
			case K_ESCPREFIX:
				rc = answer("escprefix");
				break;
		}

	} else if (!strcasecmp("GKBLED", action)) {
		if (ioctl(fd, KDGKBLED, &flags) == -1)
			kbd_error(EXIT_FAILURE, errno, _("Unable to read keyboard flags"));

		mode = (flags & 0x7);

		if (value) {
			if (((mode & LED_SCR) && !strcasecmp(value, "scrolllock")) ||
			    ((mode & LED_NUM) && !strcasecmp(value, "numlock")) ||
			    ((mode & LED_CAP) && !strcasecmp(value, "capslock")))
				rc = EXIT_SUCCESS;
		} else {
			printf("scrolllock:%s ", (mode & LED_SCR) ? "on" : "off");
			printf("numlock:%s ", (mode & LED_NUM) ? "on" : "off");
			printf("capslock:%s\n", (mode & LED_CAP) ? "on" : "off");
			rc = EXIT_SUCCESS;
		}

	} else {
		kbd_warning(0, _("Unrecognized argument: %s"), action);
	}

	close(fd);
	return rc;
}
